x-pack/plugins/alerting/server/authorization/alerting_authorization.ts

```diff
@@ -355,28 +368,42 @@ export class AlertingAuthorization {
       );
 
       // add an empty `authorizedConsumers` array on each ruleType
-      const ruleTypesWithAuthorization = this.augmentWithAuthorizedConsumers(ruleTypes, {});
-
+      const ruleTypesWithAuthorization = Array.from(
+        this.augmentWithAuthorizedConsumers(ruleTypes, {})
+      );
+      const ruleTypesAuthorized: Set<RegistryRuleType> = new Set();
       // map from privilege to ruleType which we can refer back to when analyzing the result
       // of checkPrivileges
       const privilegeToRuleType = new Map<
         string,
         [RegistryAlertTypeWithAuth, string, HasPrivileges, IsAuthorizedAtProducerLevel]
       >();
-      // as we can't ask ES for the user's individual privileges we need to ask for each feature
-      // and ruleType in the system whether this user has this privilege
-      for (const ruleType of ruleTypesWithAuthorization) {
-        for (const feature of fIds) {
-          for (const operation of operations) {
-            privilegeToRuleType.set(
-              this.authorization!.actions.alerting.get(
-                ruleType.id,
-                feature,
-                authorizationEntity,
-                operation
-              ),
-              [ruleType, feature, hasPrivilegeByOperation(operation), ruleType.producer === feature]
-            );
+      for (const feature of fIds) {
+        const featureDef = this.features
+          .getKibanaFeatures()
+          .find((kFeature) => kFeature.id === feature);
+        for (const ruleTypeId of featureDef?.alerting ?? []) {
+          const ruleTypeAuth = ruleTypesWithAuthorization.find((rtwa) => rtwa.id === ruleTypeId);
+          if (ruleTypeAuth) {
+            const { authorizedConsumers, hasAlertsMappings, hasFieldsForAAD, ...ruleType } =
+              ruleTypeAuth;
+            ruleTypesAuthorized.add(ruleType);
```

Comment:
**nit:** I did a log of this object at the end of the for loops and `.es-query` is there twice. Does this mean the set doesn't remove duplicates because they are objects? Are there any consequences if `authorizedRuleTypes` returns duplicates?

Reply:
yes that make sense, I will add some logic to avoid that!